libtiledbsoma/src/soma/soma_array.cc

```diff
@@ -1178,6 +1178,25 @@ uint64_t SOMAArray::nnz_slow() {
 }
 
 std::vector<int64_t> SOMAArray::shape() {
+    // There are two reasons for this:
+    // * Transitional, non-monolithic, phased, careful development for the
+    //   new-shape feature
+    // * Even after the new-shape feature is fully released, there will be old
+    //   arrays on disk that were created before this feature existed.
+    // So this is long-term code.
+    auto current_domain = _get_current_domain();
+    if (current_domain.is_empty()) {
+        return _tiledb_domain();
+    } else {
+        return _tiledb_current_domain();
+    }
+}
+
+std::vector<int64_t> SOMAArray::maxshape() {
+    return _tiledb_domain();
+}
+
+std::vector<int64_t> SOMAArray::_tiledb_domain() {
     std::vector<int64_t> result;
     auto dimensions = mq_->schema()->domain().dimensions();
 
```

Comment:
This is old code, but incorrect: the return value is `std::vector<int64_t>` but it accepts switch statements for other types.

Meanwhile, in the SOMA data model:

* `SparseNDArray` has dims which are all necessarily `int64_t`
* `DenseNDArray` has dims which are all necessarily `int64_t`
* `DataFrame` _default_ indexing is one dim

Reply:
Split out as #2915.